### ISSUE TITLE
Not decoding repeated field type consistently

### ISSUE THREAD

**Issue by moxley** [tomas-abrahamsson/gpb#117]
```
syntax = "proto3";

message Position {
  repeated double values = 1;
}
```

Using the above definition, and generating a file of an encoded `Position(values: [-100.0, 30.0])` from the Ruby wrapper of Google protocol buffers (`google-protobuf` gem), I get `{'Position',[]}` when decoding that file in gpb, instead of the expected `{'Position',[-100.0,30.0]}`. The same file can be decoded correctly using the Google protocol buffer C++ library. Attached is the encoded Position file generated from Ruby.

Also, when using the NIF option of gpb, the file is decoded correctly.

[position.protobuf.zip](https://github.com/tomas-abrahamsson/gpb/files/1369394/position.protobuf.zip)

**Comment by tomas-abrahamsson** [comment]
Confirmed, and I think I've found why: since the file is `proto3`, repeated doubles are `packed` by default. The packed format is: the length (a varint) followed by values, whereas a repeated field that is not packed is just the field occurring several times. The attached encoded Position file has the repeated fields in non-packed format, but the gpb currently expects only packed format for repeated fields, and I just learned this is wrong. The [doc](https://developers.google.com/protocol-buffers/docs/encoding#packed) says

> Protocol buffer parsers must be able to parse repeated fields that were compiled as packed as if they were not packed, and vice versa. [...]

I will look into fixing this. Urgent?

**Comment by moxley** [comment]
Thanks for looking into that. Yes, it's kind of urgent. We were hoping to have this work done by Oct 20. We are looking into the NIF option, but that opens a big can of worms for our deployment strategy. I wish we knew more about the internals of protobufs. We're at a loss trying to resolve this ourselves.

**Comment by tomas-abrahamsson** [comment]
I see. I will have to investigate a little bit first, to see if this is an easy or complicated change.

**Comment by tomas-abrahamsson** [comment]
It turned out to maybe not be so complicated as I had feared. I pushed to a branch, `fix-decode-packed`, in case you'd want to try it out. I'll merge it to master if you find it works well, or at the lastest in a few days, I think.

**Comment by moxley** [comment]
That works! Thank you for fixing it.

We had been investigating the NIF option, and there may come a time when we need to use it. Do you have any guidance on using the NIF option to make it work for our Elixir app? It's one thing to build it for an Erlang app, but it's another to incorporate that into a mix.exs dependency definition. Is it known to work with proto definitions that involve namespaces and imports of other proto definition files? We have such a setup, and we ran into an issue where the generated nif.cc file was referencing a doubled namespace (e.g., `Namespace::Namespace::ProtoType`). We were using the `use_packages` option in gpb_compile.

Thank you!

**Comment by tomas-abrahamsson** [comment]
I don't know enough about mix.exs to be of much use, I'm afraid. The double namespace thing rings a distant bell from long ago that it just isn't very well supported. It'd better be a separete issue though.

I'll merge to master within a few days, to see if I can fix a dialyzer issue in lrascao/rebar3_gpb_plugin#51 first.

**Comment by moxley** [comment]
Awesome. Thanks @tomas-abrahamsson!!!

**Comment by tomas-abrahamsson** [comment]
Included in 3.28.0 which I just pushed, hence closing this ticket. If there's anything, just feel free to re-open or open another ticket.

**Comment by moxley** [comment]
Thanks @tomas-abrahamsson!!!
